diff --git a/node_modules/@hotwired/turbo/dist/turbo.es2017-esm.js b/node_modules/@hotwired/turbo/dist/turbo.es2017-esm.js
index c895234..78865aa 100644
--- a/node_modules/@hotwired/turbo/dist/turbo.es2017-esm.js
+++ b/node_modules/@hotwired/turbo/dist/turbo.es2017-esm.js
@@ -3439,7 +3439,7 @@ class ProgressBar {
   constructor() {
     this.stylesheetElement = this.createStylesheetElement();
     this.progressElement = this.createProgressElement();
-    this.installStylesheetElement();
+    /**this.installStylesheetElement();**/
     this.setValue(0);
   }
 
diff --git a/node_modules/@hotwired/turbo/dist/turbo.es2017-umd.js b/node_modules/@hotwired/turbo/dist/turbo.es2017-umd.js
index 541722d..f6bebe2 100644
--- a/node_modules/@hotwired/turbo/dist/turbo.es2017-umd.js
+++ b/node_modules/@hotwired/turbo/dist/turbo.es2017-umd.js
@@ -3445,7 +3445,7 @@ Copyright © 2025 37signals LLC
     constructor() {
       this.stylesheetElement = this.createStylesheetElement();
       this.progressElement = this.createProgressElement();
-      this.installStylesheetElement();
+      /**this.installStylesheetElement();**/
       this.setValue(0);
     }
 
