// conventional-changelog-writer/dist/cli/utils.js
import { resolve, extname } from 'path';
import { pathToFileURL } from 'url';
import { readFile } from 'fs/promises';
const NEWLINE = /\r?\n/;
export async function* parseJsonStream(stream) {
    let chunk;
    let payload;
    let buffer = '';
    let json;
    for await (chunk of stream) {
        buffer += chunk.toString();
        if (NEWLINE.test(buffer)) {
            payload = buffer.split(NEWLINE);
            buffer = payload.pop() || '';
            for (json of payload) {
                try {
                    yield JSON.parse(json);
                }
                catch (err) {
                    throw new Error('Failed to split commits', {
                        cause: err
                    });
                }
            }
        }
    }
    if (buffer) {
        try {
            yield JSON.parse(buffer);
        }
        catch (err) {
            throw new Error('Failed to split commits', {
                cause: err
            });
        }
    }
}
export async function* readCommitsFromFiles(files) {
    for (const file of files) {
        try {
            yield JSON.parse(await readFile(file, 'utf8'));
        }
        catch (err) {
            console.warn(`Failed to read file ${file}:\n  ${err}`);
        }
    }
}
export function readCommitsFromStdin() {
    return parseJsonStream(process.stdin);
}
function relativeResolve(filePath) {
    return pathToFileURL(resolve(process.cwd(), filePath));
}
export async function loadDataFile(filePath) {
    const resolvedFilePath = relativeResolve(filePath);
    const ext = extname(resolvedFilePath.toString());
    if (ext === '.json') {
        return JSON.parse(await readFile(resolvedFilePath, 'utf8'));
    }
    // @ts-expect-error Dynamic import actually works with file URLs
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    return (await import(resolvedFilePath)).default;
}
