// conventional-commits-parser/dist/cli/utils.js
import fs from 'fs';
import readline from 'readline';
export async function* splitStream(stream, separator) {
    let chunk;
    let payload;
    let buffer = '';
    for await (chunk of stream) {
        buffer += chunk.toString();
        if (buffer.includes(separator)) {
            payload = buffer.split(separator);
            buffer = payload.pop() || '';
            yield* payload;
        }
    }
    if (buffer) {
        yield buffer;
    }
}
export async function* readRawCommitsFromFiles(files, separator) {
    for (const file of files) {
        try {
            yield* splitStream(fs.createReadStream(file), separator);
        }
        catch (err) {
            console.warn(`Failed to read file ${file}:\n  ${err}`);
        }
    }
}
export async function* readRawCommitsFromLine(separator) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: true
    });
    let line = '';
    let commit = '';
    for await (line of rl) {
        commit += `${line}\n`;
        if (!commit.includes(separator)) {
            return;
        }
        yield commit;
        commit = '';
    }
}
export function readRawCommitsFromStdin(separator) {
    return splitStream(process.stdin, separator);
}
const JSON_STREAM_OPEN = '[\n';
const JSON_STREAM_SEPARATOR = '\n,\n';
const JSON_STREAM_CLOSE = '\n]\n';
export async function* stringify(commits) {
    let jsonStreamOpened = false;
    yield JSON_STREAM_OPEN;
    for await (const commit of commits) {
        if (jsonStreamOpened) {
            yield JSON_STREAM_SEPARATOR;
        }
        yield JSON.stringify(commit);
        jsonStreamOpened = true;
    }
    yield JSON_STREAM_CLOSE;
}
